File: src/algorithms/hypercube.cpp
#include "../../include/algorithms/hypercube.hpp"
#include <random>
#include <algorithm>
#include <numeric>
#include <unordered_set>
#include <chrono>
#include <iostream>
#include <vector>

/**
 * Builds the Hypercube index for the input dataset ref 24
 * @param data Input dataset to be indexed
 */
void Hypercube::buildIndex()
{
    // slide 24: d' = floor(log_2 n) - {1,2,3}
    const size_t n = 
        Data.size();
    int dlog = (n > 0) ? (int)floor(log2((double)max<size_t>(1, n))) : 1;
    kproj_ = (Args.kproj > 0) ? Args.kproj : max(1, dlog - 2);
    Dim = Data[0].values.size();

    // slide 18: w ∈ [2, 6], larger for range queries
    w_ = (Args.w > 0) ? Args.w : 4.0f;

    // Random number generators
    mt19937_64 rng(Args.seed);
    normal_distribution<double> normal(0.0, 1.0);    // For LSH projections
    uniform_real_distribution<float> unif(0.0f, w_); // For LSH shifts
    uniform_int_distribution<uint32_t> uni32(1u, 0xffffffffu); // For f_j hash functions

    // slide 24: Generate d' base LSH functions h_j
    // Each h_j(p) = floor((a_j · p + t_j) / )
    proj_.assign(kproj_, vector<float>(Dim));
    shift_.assign(kproj_, 0.0f);
    for (int j = 0; j < kproj_; ++j)
    {
        for (int d = 0; d < Dim; ++d)
            proj_[j][d] = (float)normal(rng);
        shift_[j] = unif(rng);
    }

    // slide 24: Generate bit-mapping functions f_j: Z → {0,1}
    // f_j maps buckets to bits uniformly at random
    // Implementation: use random hash function f_j(h) = (aₕ · h + bₕ) mod 2
    fA_.resize(kproj_);
    fB_.resize(kproj_);
    for (int j = 0; j < kproj_; ++j)
    {
        fA_[j] = (uint64_t)uni32(rng);
        fB_[j] = (uint64_t)uni32(rng);
        if (fA_[j] == 0)
            fA_[j] = 1; // Ensure non-zero
    }

    // slide 25: Choose storage strategy based on d'
    // Dense array for small d' (≤ 24 bits = 16M vertices)
    // Sparse map for large d'
    denseCube_ = (kproj_ <= 24);
    if (denseCube_)
    {
        cubeDense_.assign(1ULL << kproj_, {});
        cubeSparse_.clear();
    }
    else
    {
        cubeSparse_.clear();
        cubeDense_.clear();
    }

    // slide 24: Insert all data points into their vertices
    for (int id = 0; id < (int)n; ++id)
    {
        const auto &v = Data[id].values;
        uint64_t vtx = vertexOf(v);
        if (denseCube_)
            cubeDense_[vtx].push_back(id);
        else
            cubeSparse_[vtx].push_back(id);
    }

    if (getenv("CUBE_DEBUG"))
    {
        cerr << "[CUBE] n=" << n
                  << " dim=" << Dim
                  << " d'=" << kproj_
                  << " w=" << w_
                  << " dense=" << denseCube_
                  << " vertices_allocated=" << (denseCube_ ? (1ULL << kproj_) : cubeSparse_.size())
                  << "\n";
    }
}

/**
 * Computes h_j(v) = floor((a_j · v + t_j) / w) for projection j ref 18
 * @param v Input vector
 * @param j Projection index
 * @return Hash value
 */
long long Hypercube::hij(const vector<float> &v, int j) const
{
    double dot = 0.0;
    for (int d = 0; d < Dim; ++d)
        dot += proj_[j][d] * v[d];
    return (long long)floor((dot + shift_[j]) / w_);
}

/**
 * Maps hash value to bit using function f_j ref 24
 * @param h Hash value
 * @param j Projection index
 * @return Bit value (0 or 1)
 */
bool Hypercube::fj(long long h, int j) const
{
    // Random hash function: (a·h + b) mod 2
    uint64_t hval = (uint64_t)(h >= 0 ? h : -h);
    return ((fA_[j] * hval + fB_[j]) & 1) != 0;
}

/**
 * Maps a vector to a vertex in the hypercube ref 24
 * Vertex label is d'-bit string where bit j = f_j(h_j(v))
 * @param v Input vector
 * @return 64-bit vertex label
 */
uint64_t Hypercube::vertexOf(const vector<float> &v) const
{
    uint64_t key = 0;
    for (int j = 0; j < kproj_; ++j)
    {
        long long hj = hij(v, j);    // Compute LSH hash
        if (fj(hj, j))               // Map to bit using f_j
            key |= (1ULL << j);
    }
    return key;
}

/**
 * Generates vertices to probe in order of increasing Hamming distance ref 24
 * @param base Starting vertex (query's vertex)
 * @param kproj Number of bits (d')
 * @param maxProbes Maximum number of vertices to return
 * @param maxHamming Maximum Hamming distance to probe ref 25
 * @return Vector of vertex labels to probe
 */
vector<uint64_t>
Hypercube::probesList(uint64_t base, int kproj, int maxProbes, int maxHamming) const
{
    vector<uint64_t> out{base};
    if ((int)out.size() >= maxProbes)
        return out;

    // slide 25: Threshold on Hamming distance
    const int Hmax = min(kproj, maxHamming);
    
    // Generate vertices at increasing Hamming distances: 1, 2, ...
    for (int h = 1; h <= Hmax && (int)out.size() < maxProbes; ++h)
    {
        // Generate all combinations of h bit flips
        vector<int> idx(h);
        iota(idx.begin(), idx.end(), 0);
        
        while (true)
        {
            // Flip bits at positions in idx
            uint64_t mask = 0;
            for (int i : idx)
                mask |= (1ULL << i);
            out.push_back(base ^ mask);
            
            if ((int)out.size() >= maxProbes)
                break;
            
            // Next combination in lexicographic order
            int i;
            for (i = h - 1; i >= 0 && idx[i] == i + kproj - h; --i)
                ;
            if (i < 0)
                break;
            ++idx[i];
            for (int j = i + 1; j < h; ++j)
                idx[j] = idx[j - 1] + 1;
        }
    }
    return out;
}

/**
 * Performs Hypercube search for all queries 
 * @param queries Query dataset
 * @param out Output file stream for results
 */
void Hypercube::search(const vector<VectorData> &queries, ofstream &out)
{
    using namespace chrono;
    out << "Hypercube\n\n";

    // slide 25: Search parameters
    int M = Args.M;           // Threshold: max candidates to check in R^d
    int probes = Args.probes; // Threshold: max vertices to probe
    int maxHam = (Args.maxHamming > 0) ? Args.maxHamming : kproj_; // Hamming distance bound
    double R = Args.R;
    bool doRange = Args.rangeSearch;
    int N = Args.N;

    int Q = (int)queries.size();
    
    if (Args.maxQueries > 0)
        Q = min(Q, static_cast<int>(Args.maxQueries));

    for (int qi = 0; qi < Q; ++qi)
    {
        const auto &q = queries[qi].values;
        auto t0 = high_resolution_clock::now();

        // slide 24 Project query to hypercube vertex
        uint64_t base = vertexOf(q);
        
        // slide 24 Check points in same and nearby vertices
        // Generate vertices in increasing Hamming distance
        auto plist = probesList(base, kproj_, probes, maxHam);
        
        // Collect unique candidates from probed vertices
        unordered_set<int> candSet;
        candSet.reserve(min<size_t>((size_t)M, (size_t)4096));
        size_t gathered = 0;
        
        for (auto vtx : plist)
        {
            const vector<int> *bucket = nullptr;
            if (denseCube_)
            {
                bucket = &cubeDense_[vtx];
            }
            else
            {
                auto it = cubeSparse_.find(vtx);
                if (it == cubeSparse_.end())
                    continue;
                bucket = &it->second;
            }
            
            // Add points from this vertex
            for (int id : *bucket)
            {
                if (candSet.insert(id).second && ++gathered >= (size_t)M)
                    break;
            }
            if (gathered >= (size_t)M)
                break;
        }

        // Compute actual distances to candidates
        vector<pair<double, int>> distApprox;
        distApprox.reserve(candSet.size());
        vector<int> rlist;

        for (int id : candSet)
        {
            double d = l2(q, 
                          Data[id].values);
            distApprox.emplace_back(d, id);
            if (doRange && d <= R)
                rlist.push_back(id);
        }

        // slide 24 Return closest candidates or range search results
        int topN = min(N, (int)distApprox.size());
        if (topN > 0)
        {
            nth_element(distApprox.begin(), distApprox.begin() + topN, distApprox.end());
            sort(distApprox.begin(), distApprox.begin() + topN);
            distApprox.resize(topN);
        }
        
        double tApprox = duration<double>(high_resolution_clock::now() - t0).count();
        calculatePerQueryMetrics(queries[qi].id, qi, tApprox, distApprox, rlist, out);
    }
    printSummary(Q, out);
}

File: prj3_files/methods/hypercube.py
import numpy as np
from typing import List, Tuple, Dict
from collections import defaultdict
from itertools import combinations


class Hypercube:
    def __init__(
        self,
        kproj: int = None,
        w: float = 4.0,
        max_probes: int = 100,
        max_hamming: int = 3,
        M: int = 1000,
        metric: str = 'L2',
        seed: int = 42
    ):
        self.kproj = kproj
        self.w = w
        self.max_probes = max_probes
        self.max_hamming = max_hamming
        self.M = M
        self.metric = metric.lower()
        self.seed = seed
        
        self.dim = None
        self.n = None
        self.proj_ = None
        self.shift_ = None
        self.fA_ = None
        self.fB_ = None
        self.cube_ = None
        self.embeddings = None
        
        self.rng = np.random.RandomState(seed)
    
    def _normalize(self, vectors: np.ndarray) -> np.ndarray:
        if vectors.ndim == 1:
            norm = np.linalg.norm(vectors)
            return vectors / max(norm, 1e-8)
        norms = np.linalg.norm(vectors, axis=1, keepdims=True)
        norms = np.maximum(norms, 1e-8)
        return vectors / norms
    
    def build_index(self, embeddings: np.ndarray):
        self.n, self.dim = embeddings.shape
        self.embeddings = embeddings.astype(np.float32)
        
        # Normalize for cosine similarity
        if self.metric == 'cosine':
            self.embeddings = self._normalize(self.embeddings)
        
        if self.kproj is None:
            dlog = int(np.floor(np.log2(max(1, self.n)))) if self.n > 0 else 1
            self.kproj = max(1, dlog - 2)
        self.kproj = min(self.kproj, 24)
        
        print(f"Building Hypercube index:")
        print(f"  N = {self.n} proteins")
        print(f"  D = {self.dim} dimensions")
        print(f"  d' = {self.kproj} (hypercube dimension)")
        print(f"  metric = {self.metric}")
        if self.metric == 'l2':
            print(f"  w = {self.w}")
        print(f"  max_probes = {self.max_probes}")
        print(f"  max_hamming = {self.max_hamming}")
        
        self._generate_hash_functions()
        
        self.cube_ = defaultdict(list)
        
        print("  Inserting points...")
        for idx in range(self.n):
            vec = self.embeddings[idx]
            vertex = self._vertex_of(vec)
            self.cube_[vertex].append(idx)
        
        print(f"Hypercube index built!")
        print(f"Total vertices occupied: {len(self.cube_)}")
    
    def _generate_hash_functions(self):
        self.proj_ = self.rng.randn(self.kproj, self.dim).astype(np.float32)
        
        if self.metric == 'l2':
            self.shift_ = self.rng.uniform(0, self.w, size=self.kproj).astype(np.float32)
        
        self.fA_ = self.rng.randint(1, 2**32 - 1, size=self.kproj, dtype=np.uint64)
        self.fB_ = self.rng.randint(0, 2**32 - 1, size=self.kproj, dtype=np.uint64)
    
    def _h_ij(self, v: np.ndarray, j: int) -> int:
        dot = np.dot(self.proj_[j], v)
        if self.metric == 'cosine':
            # For cosine: use sign of dot product
            return 1 if dot >= 0 else 0
        else:
            return int(np.floor((dot + self.shift_[j]) / self.w))
    
    def _f_j(self, h: int, j: int) -> bool:
        if self.metric == 'cosine':
            # For cosine: h is already 0 or 1
            return h == 1
        else:
            h_val = abs(h)
            return ((self.fA_[j] * h_val + self.fB_[j]) & 1) != 0
    
    def _vertex_of(self, v: np.ndarray) -> int:
        key = 0
        for j in range(self.kproj):
            h_j = self._h_ij(v, j)
            if self._f_j(h_j, j):
                key |= (1 << j)
        return key
    
    def _probes_list(self, base: int) -> List[int]:
        probes = [base]
        
        if len(probes) >= self.max_probes:
            return probes
        
        H_max = min(self.kproj, self.max_hamming)
        
        for h in range(1, H_max + 1):
            if len(probes) >= self.max_probes:
                break
            
            for positions in combinations(range(self.kproj), h):
                mask = 0
                for pos in positions:
                    mask |= (1 << pos)
                
                probes.append(base ^ mask)
                
                if len(probes) >= self.max_probes:
                    break
        
        return probes
    
    def _compute_distance(self, v1: np.ndarray, v2: np.ndarray) -> float:
        if self.metric == 'cosine':
            sim = np.dot(v1, v2) / (np.linalg.norm(v1) * np.linalg.norm(v2) + 1e-8)
            return 1.0 - sim
        else:
            return np.linalg.norm(v1 - v2)
    
    def search(
        self,
        query: np.ndarray,
        N: int = 10
    ) -> List[Tuple[int, float]]:
        
        # Normalize query for cosine similarity
        if self.metric == 'cosine':
            query = self._normalize(query)
        
        base = self._vertex_of(query)
        probe_list = self._probes_list(base)
        candidates = set()
        gathered = 0
        
        for vertex in probe_list:
            if vertex not in self.cube_:
                continue
            
            for idx in self.cube_[vertex]:
                if idx not in candidates:
                    candidates.add(idx)
                    gathered += 1
                    
                    if gathered >= self.M:
                        break
            
            if gathered >= self.M:
                break
        
        distances = []
        for idx in candidates:
            dist = self._compute_distance(query, self.embeddings[idx])
            distances.append((idx, dist))
        
        distances.sort(key=lambda x: x[1])
        return distances[:N]
    
    def batch_search(
        self,
        queries: np.ndarray,
        N: int = 10,
        verbose: bool = True
    ) -> List[List[Tuple[int, float]]]:
        
        results = []
        num_queries = len(queries)
        
        if verbose:
            from tqdm import tqdm
            iterator = tqdm(range(num_queries), desc="Hypercube search")
        else:
            iterator = range(num_queries)
        
        for qi in iterator:
            result = self.search(queries[qi], N=N)
            results.append(result)
        
        return results